openedx/core/djangoapps/waffle_utils/__init__.py

```diff
@@ -291,8 +293,60 @@ def is_flag_active(self, flag_name, check_before_waffle_callback=None, flag_unde
                         return bool(flag_undefined_default)
 
                 self._cached_flags[namespaced_flag_name] = value
+
+        self._set_waffle_flag_metric(namespaced_flag_name, value)
         return value
 
+    def _set_waffle_flag_metric(self, name, value):
+        """
+        If ENABLE_WAFFLE_FLAG_METRIC is True, adds name/value to cached values
+        and sets metric if the value changed.
+
+        Metric flag values could be False, True, or Both. The value Both would
+        mean that the flag had both a True and False value at different times
+        through the transaction. This is most likely due to having a
+        check_before_waffle_callback, as is the case with CourseWaffleFlag.
+
+        Example metric value::
+
+            "{'another.course.flag': 'False', 'some.flag': 'False', 'some.course.flag': 'Both'}"
+
+        Warning: NewRelic does not recommend large custom metric values due to
+        the potential performance impact on the agent, so you may just want to
+        enable when researching usage of a particular flag. Metric values longer
+        than 255 are truncated.
+
+        """
+        is_waffle_flag_metric_enabled = getattr(settings, _ENABLE_WAFFLE_FLAG_METRIC, False)
+        if not is_waffle_flag_metric_enabled:
+            return
+
+        flag_metric_data = self._get_request_cache().setdefault('flag_metric', {})
+        is_value_change = False
+        if name not in flag_metric_data:
+            flag_metric_data[name] = str(value)
+            is_value_change = True
+        else:
+            if flag_metric_data[name] != str(value):
+                flag_metric_data[name] = 'Both'
+                is_value_change = True
+
+        if is_value_change:
+            set_custom_metric('waffle_flags', str(flag_metric_data))
```

Comment:
Future: we can consider adding a capability to [MonitoringCustomMetricsMiddleware](https://github.com/edx/edx-django-utils/blob/196966d05f0b437ec02ad7c233335491874c20cd/edx_django_utils/monitoring/middleware.py#L34) to call NewRelic agent only once at the end of the transaction.  Note that NewRelic only sends data to its server every 60 seconds, so it would reduce calls/processing by the agent, but should have minimal impact on the data sent to the server (unless the transaction happens to cross the 60 second boundary).